Modify 'link' plugin dialog to allow creation of links to 'autoid' headings

sbaughman opened the pull request:
This adds the ability to add links in a document that target `autoid` headings.

It modifies the `link` plugin, specifically the dialog window definition, in order to add the necessary UI elements and modify functions to allow the additional data to be processed correctly

sbaughman committed:
Set up listener for link dialog definition

sbaughman committed:
Extract event callback function

sbaughman committed:
Add new option to linkTypes dropdown

sbaughman committed:
Modify onChange function for link type

sbaughman committed:
Add explanation and reference to 'link' plugin on github

sbaughman committed:
Add headingOptions to partIds

sbaughman committed:
Add some comments

sbaughman committed:
Start to add additional content definition for creating headings

sbaughman committed:
Add main field for content and a text warning if there are no headings

sbaughman committed:
Change fieldset to a select box and fix widths

sbaughman committed:
Fix alignment issues

sbaughman committed:
Add headings and heading id values to the select dropdown

sbaughman committed:
Commit the data

sbaughman committed:
Add modified 'onOk' function

sbaughman committed:
Define 'plugin' variable to avoid error when 'plugin.getLinkAttributes' is called

sbaughman committed:
Set anchor id from heading id

caffodian commented on plugins/autoid/plugin.js:
```diff
@@ -42,6 +42,8 @@
         }
       });
 
+      CKEDITOR.on('dialogDefinition', modifyLinkDialog);
```
Interesting!  This only fires once.

sbaughman committed:
Add if check to make sure data type does not always get set to anchor

caffodian commented on plugins/autoid/plugin.js:
```diff
@@ -158,6 +160,208 @@
+        }
+      }
+
+      function modifiedLinkTypeChanged() {
```
This looks very much like https://github.com/ckeditor/ckeditor-dev/blob/master/plugins/link/dialogs/link.js#L46 with an extra option added (because it is).

I was going to suggest trying to do this in a way where we just have a separate function that shows the new tab + calls the old function, but my 5 minute glance says I don't think this is possible without heavy refactoring of the original dialog.

Did you look into this as well?

sbaughman replied:
Yes.  Actually this function would not be needed at all if the heading link options were in an entirely separate tab.  I do, however, think that that would make for a less intuitive user interface, since all the other link types are accessed through a single tab.

caffodian commented on plugins/autoid/plugin.js:
```diff
@@ -158,6 +160,209 @@
+            type: 'html',
+            id: 'noHeadings',
+            style: 'text-align: center;',
+            html: '<div role="note" tabIndex="-1">' + CKEDITOR.tools.htmlEncode( '(No headings available in the document)' ) + '</div>',
```
Maybe "No headings created yet in the document" ?

sbaughman replied:
I can change it to that.  I was just mirroring what they used for the similar anchor message.

caffodian replied:
Ah.  We've never really found a lot of their messages to be very user friendly...:/

sbaughman replied:
changed

caffodian commented on plugins/autoid/plugin.js:
```diff
@@ -158,6 +160,209 @@
+        // property so it will create the new link without having to
+        // modify the plugin further
+        if (data.heading.id && !data.anchor.id)
+          data.anchor.id = data.heading.id;
```
This is a bit challenging.  I see this part is different.  Is anything else different?  

It's too bad there's not a better way to hook into the "after commit"

sbaughman replied:
This and the `plugin` variable assignment on line 281 are the only changes to this function.  Unfortunately, it's a huge function, and also the only place where I can intercept that data object and make the necessary modifications to it without having to modify the main plugin more extensively.  The other option would be the `plugin.getLinkAttributes` function, which is far larger than this one.

caffodian commented:
Could you please screenshot it in a comment?  @sbaughman 

thanks!

caffodian commented:
One further thing:  can we repurpose/crib from these tests just to have tests for the id/heading content mapping?

https://github.com/ckeditor/ckeditor-dev/blob/master/tests/plugins/link/dialog.js

sbaughman commented:
![screen shot 2016-09-29 at 5 12 37 pm](https://cloud.githubusercontent.com/assets/13240692/18972647/0914d8e2-8668-11e6-9a78-9832c4541284.png)

sbaughman committed:
Add file for heading_link tests and configure bender.js

sbaughman committed:
Copy dialog test for link plugin and add attribution

sbaughman committed:
Modify anchor discovery function for heading ids and text content

sbaughman committed:
Change startup data to use headings

sbaughman committed:
Remove 'name' attribute from headings

sbaughman committed:
Add autoid plugin and allow heading ids

Not sure if the extraAllowedContent declaration is necessary, since it
is allowed by loading the autoid plugin.  I’ll get it working, then try
removing it later on

sbaughman committed:
Change test names and assertions

sbaughman committed:
Add symlink to link plugin

sbaughman committed:
Add function to find 'linkType' tab

It is not always the second element if certain options are enabled /
disabled

sbaughman committed:
Get the correct id values and text for each option

sbaughman committed:
Correct assertion message.  Add symlink to divarea plugin

sbaughman committed:
Check to see if the heading type has already been added to link definition

Prevents adding content multiple times if there is more than one editor
on the page

sbaughman committed:
Pass the editor explicitly to the 'findAllHeadings' function

Makes sure the correct headings populate the dialog when there is more
than one editor instance on the page

sbaughman committed:
Fix inline test expectation

caffodian commented on plugins/autoid/plugin.js:
```diff
@@ -62,7 +62,7 @@
 
-      function findAllHeadings() {
+      function findAllHeadings(editor) {
         return editor.document.find('h1, h2, h3, h4, h5, h6');
```
try editor.editable() instead of editor.document?  document actually means the whole HTML document (so either the iframe, ok for iframe mode, or the entire page that the inline one is on (not so good) )

sbaughman replied:
Yup, that was it!

sbaughman committed:
Change 'findAllHeadings' to search only 'editor.editable' instead of the whole document

sbaughman committed:
Fix spacing

sbaughman committed:
Replace pasting.js with symlink

sbaughman committed:
Add 'autoid-' prefixes to unique ids added by plugin

sbaughman committed:
Add onShow function so we can modify it

sbaughman committed:
Define plugin variable

sbaughman committed:
Set data type to 'heading' if the anchor id was generated by autoid

sbaughman committed:
Fix spacing

sbaughman committed:
Set value of initialLinkText on show

sbaughman committed:
Change variable name for editor referenced in the dialog modifications

sbaughman committed:
Populate the heading selector when editing an existing link to a heading

sbaughman committed:
fix symlinks

sbaughman committed:
fix symlink

sbaughman committed:
Update ckeditor version

sbaughman committed:
Change wording
